docs/index.rst

```diff
@@ -1,260 +1,427 @@
-safe to assume after this call that ``int_value`` has the right type. If
-the key doesn’t exist in any configuration file, Confit will raise a
-``NotFoundError``. Together, catching these exceptions (both subclasses
-of ``confit.ConfigError``) lets you painlessly validate the user’s
-configuration as you go.
+Hello World
+^^^^^^^^^^^
 
+::
 
-View Theory
+  1| import confit
+  2| default_config = confit.from_mapping({'greeting': 'Hello, world!'})
+  3| sys_config = confit.from_yaml('/etc/example.yml')
+  4| config = default_config + sys_config
+  5| config['greeting'].get(str)
+
+Let's step through each line in this example:
+
+1. Import Confit.
+2. Create a configuration from code---that is, from a mapping literal. It
+   will serve as our default.
+3. Read a configuration from a `YAML`_ file in the system configuration
+   directory.
+4. Compose the two configurations by laying the system configuration over the
+   default configuration. Every variable in the system configuration will
+   override the ones in the default.
+5. Read a string variable from the configuration.
+
+.. _YAML: http://yaml.org/
+
+This example is just enough to get a taste for Confit. Details are explained
+in later sections.
+
+
+Git
+^^^
+
+To demonstrate the full power of Confit, consider the `configuration model
+for Git <http://git-scm.com/docs/git-config#FILES>`_.
+
+::
+
+  from_git_ini = lambda path: confit.from_ini(path, sep=".")
+  sys_config = from_git_ini(confit.sys_config_path("git", "config"))
+  second_user_config = sum(map(from_git_ini, confit.user_config_paths("git", "config")))
+  first_user_config = from_git_ini(os.expanduser("~/.gitconfig"))
+  repo_config = from_git_ini(os.path.join(git_dir(), "config"))
+  # Git takes configuration from environment variables with names different
+  # from the configuration variables they map to. That requires some custom
+  # (but simple) logic not built in to Confit. It is left as an exercise for
+  # the reader.
+  env_config = ...
+  config = sum((defaults, sys_config, second_user_config, first_user_config, repo_config, env_config))
+
+
+curl
+^^^^
+
+Confit is useful for libraries in addition to applications. Consider libcurl,
+which loads configuration from `~/.curlrc` and special environment variables
+(`http_proxy`, `https_proxy`, `no_proxy`, etc). Libraries can use all the
+facilities of Confit (with the exception of ``from_args``) to ease handling
+of their configuration.
+
+
+Configuration
+-------------
+
+A configuration consists of variable bindings. In each new project, a
+configuration is born when some variable holds an arbitrary value, e.g.  a
+path to a temporary directory. The value is moved to the configuration as a
+default, and the variable's value is loaded from the configuration.
+
+Variables
+^^^^^^^^^
+
+In Confit, a configuration is an immutable mapping from string names to
+*variables*. Variables have two methods, ``get`` and ``origin``.
+
+::
+
+  get(cast=lambda x: x)
```

Comment:
This is worth breaking out into a separate discussion: The largest issue I run into with the current Confit is the cumbersome need to re-validate the same configuration option in different parts of the code. It can be error-prone and verbose to need to copy and paste `config['something'].get(some_complicated_validator)` all over. This can even get to be a performance problem for really complicated conversions.

I don't actually have a solution to this. The current template system is a partial solution, but we may be able to do better. Bears thinking about more.